contracts/modules/cover/Cover.sol

```diff
@@ -0,0 +1,316 @@
+
+      if (remainder > 0) {
+        // solhint-disable-next-line avoid-low-level-calls
+        (bool ok, /* data */) = address(msg.sender).call{value: remainder}("");
+        require(ok, "Cover: Returning ETH remainder to sender failed.");
+      }
+    } else {
+      IERC20 token = IERC20(payoutAsset);
+      token.transferFrom(msg.sender, address(this), totalPrice);
+    }
+  }
+
+  /* ========== PRICE CALCULATION ========== */
+
+  function getPrice(uint amount, uint period, uint productId, IStakingPool pool) public view returns (uint, uint) {
+    uint basePrice = interpolatePrice(
+      pool.getStake(productId),
+      lastPrices[productId][address(pool)] != 0 ? lastPrices[productId][address(pool)] : initialPrices[productId],
+      pool.getTargetPrice(productId),
+      lastPriceUpdate[productId][address(pool)],
+      block.timestamp
+    );
+    uint pricePercentage = calculatePrice(
+      amount,
+      basePrice,
+      pool.getUsedCapacity(productId),
+      pool.getCapacity(productId, capacityFactors[productId])
+    );
+
+    uint price = pricePercentage * amount * period / 365 days / MAX_PRICE_PERCENTAGE;
+    return (pricePercentage, price);
+  }
+
+  /**
+    Price changes towards targetPrice from lastPrice by maximum of 1% a day per every 100k NXM staked
+  */
+  function interpolatePrice(
+    uint stakedNXM,
+    uint lastPrice,
+    uint targetPrice,
+    uint lastPriceUpdate,
+    uint now
+  ) public pure returns (uint) {
+
+    uint percentageChange = (now - lastPriceUpdate) / 1 days  * (stakedNXM / STAKE_SPEED_UNIT) * PERCENTAGE_CHANGE_PER_DAY_BPS;
+    if (targetPrice > lastPrice) {
+      return lastPrice + (targetPrice - lastPrice) * percentageChange / BASIS_PRECISION;
+    } else {
+      return lastPrice - (lastPrice - targetPrice) * percentageChange / BASIS_PRECISION;
+    }
+  }
+
+  function calculatePrice(
+    uint amount,
+    uint basePrice,
+    uint activeCover,
+    uint capacity
+  ) public pure returns (uint) {
+    return (calculatePriceIntegralAtPoint(
+      basePrice,
+      activeCover + amount,
+      capacity
+    ) -
+    calculatePriceIntegralAtPoint(
+      basePrice,
+      activeCover,
+      capacity
+    )) / amount;
+  }
+
+  function calculatePriceIntegralAtPoint(
+    uint basePrice,
+    uint activeCover,
+    uint capacity
+  ) public pure returns (uint) {
+    uint actualPrice = basePrice * activeCover;
+    for (uint i = 0; i < PRICE_CURVE_EXPONENT; i++) {
+      actualPrice = actualPrice * activeCover / capacity;
+    }
+    actualPrice = actualPrice / 8 + basePrice * activeCover;
+
+    return actualPrice;
+  }
+
+  /* ========== PRODUCT CONFIGURATION ========== */
+
+  function setCapacityFactor(uint productId, uint capacityFactor) external onlyAdvisoryBoard {
+    capacityFactors[productId] = capacityFactor;
+  }
+
+  function setInitialPrice(uint productId, uint initialPrice) external onlyAdvisoryBoard {
+    initialPrices[productId] = initialPrice;
+  }
```

Comment:
we also need a function to list new products, where you configure all settings in one tx